source/DoublyLinkedList.h

```diff
@@ -0,0 +1,274 @@
+#ifndef __DATA_STRUCTURES_DOUBLY_LINKED_LIST
+#define __DATA_STRUCTURES_DOUBLY_LINKED_LIST
+
+#include <functional>
+#include <iterator>
+#include <type_traits>
+
+namespace DataStructures {
+
+template<typename T>
+class DoublyLinkedList
+{
+public:
+  static_assert(std::is_same<decltype(T{} == T{}), bool>(),
+                "value type must have `operator==(T&)` defined");
+  static_assert(std::is_same<decltype(T{} != T{}), bool>(),
+                "value type must have `operator!=(T&)` defined");
+
+private:
+  struct Element
+  {
+    T datum;
+    Element* previous;
+    Element* next;
+  };
+
+  Element* first;
+  Element* last;
+  size_t number_of_elements;
+
+public:
+  /**
+    A type for iterating forward through the list.
+  */
+  class iterator : std::iterator<std::forward_iterator_tag, T>
```

Comment:
For a doubly linked list, check out [std::bidirectional_iterator](https://en.cppreference.com/w/cpp/iterator/bidirectional_iterator).